contracts/instruments/RibbonThetaVaultYearn.sol

```diff
@@ -0,0 +1,815 @@
+            wmul(amount, collateralToken.pricePerShare());
+        _deposit(collateralToAssetBalance);
+    }
+
+    /**
+     * @notice Mints the vault shares to the msg.sender
+     * @param amount is the amount of `asset` deposited
+     */
+    function _deposit(uint256 amount) private {
+        uint256 totalWithDepositedAmount = totalBalance();
+        require(totalWithDepositedAmount < cap, "Cap exceeded");
+        require(
+            totalWithDepositedAmount >= MINIMUM_SUPPLY,
+            "Insufficient asset balance"
+        );
+
+        // amount needs to be subtracted from totalBalance because it has already been
+        // added to it from either IWETH.deposit and IERC20.safeTransferFrom
+        uint256 total = totalWithDepositedAmount.sub(amount);
+
+        uint256 shareSupply = totalSupply();
+
+        // Following the pool share calculation from Alpha Homora:
+        // solhint-disable-next-line
+        // https://github.com/AlphaFinanceLab/alphahomora/blob/340653c8ac1e9b4f23d5b81e61307bf7d02a26e8/contracts/5/Bank.sol#L104
+        uint256 share =
+            shareSupply == 0 ? amount : amount.mul(shareSupply).div(total);
+
+        require(
+            shareSupply.add(share) >= MINIMUM_SUPPLY,
+            "Insufficient share supply"
+        );
+
+        emit Deposit(msg.sender, amount, share);
+
+        _mint(msg.sender, share);
+    }
+
+    /**
+     * @notice Withdraws ETH from vault using vault shares
+     * @param share is the number of vault shares to be burned
+     */
+    function withdrawETH(uint256 share) external nonReentrant {
+        require(asset == WETH, "!WETH");
+        uint256 withdrawAmount = _withdraw(share, false, true);
+
+        IWETH(WETH).withdraw(withdrawAmount);
+        (bool success, ) = msg.sender.call{value: withdrawAmount}("");
+        require(success, "ETH transfer failed");
+    }
+
+    /**
+     * @notice Withdraws WETH from vault using vault shares
+     * @param share is the number of vault shares to be burned
+     */
+    function withdraw(uint256 share) external nonReentrant {
+        uint256 withdrawAmount = _withdraw(share, false, true);
+        IERC20(asset).safeTransfer(msg.sender, withdrawAmount);
+    }
+
+    /**
+     * @notice Withdraws yvWETH + WETH (if necessary) from vault using vault shares
+     * @param share is the number of vault shares to be burned
+     */
+    function withdrawYieldToken(uint256 share) external nonReentrant {
+        uint256 pricePerYearnShare = collateralToken.pricePerShare();
+        uint256 withdrawAmount =
+            wdiv(_withdraw(share, false, false), pricePerYearnShare);
+
+        (uint256 yieldTokenBalance, uint256 yieldTokensToWithdraw) =
+            _withdrawYieldToken(withdrawAmount, pricePerYearnShare);
+
+        // If there is not enough yvWETH in the vault, it withdraws as much as possible and
+        // transfers the rest in `asset`
+        if (yieldTokensToWithdraw == yieldTokenBalance) {
```

Comment:
Do we want to do `yieldTokensToWithdraw >= yieldTokenBalance` to be defensive?

Reply:
well, yieldTokensToWithdraw can only either be [yieldTokenBalance OR withdrawAmount ](https://github.com/ribbon-finance/structured-products/blob/7f27cf58236f43cdc532b0800347d504ee11799e/contracts/instruments/RibbonThetaVaultYearn.sol#L339)